### ISSUE TITLE
GraphiteWriter - pickle protocol support

### ISSUE THREAD

**Issue by testuser7** [jmxtrans/jmxtrans#322]
Hi,

Pickle protocol support may be a nice feature for users that collecting many counters very often.
I'm getting performance problem using InfluxDB with graphite plugin.

Regards,

**Comment by gehel** [comment]
Yep, that could be a good idea. We just need to find a good Java implementation of Pickle and hope that the performance gain we see on the Graphite / InfluxDB side is more than what we loose on the JmxTrans side ...

Pull request is welcomed ...

**Comment by testuser7** [comment]
There is GraphitePickleWriter class in embedded-jmxtrans...

**Comment by gehel** [comment]
Yep, just need to port it over ...

I'm surprised that @cyrille-leclerc added a dependency to Jython, seems like a large addition even for an optional dependency.

If anyone want to send the pull request, I'll be happy to review and merge ...

**Comment by testuser7** [comment]
Other pickle protocol java implementations:
- https://github.com/apache/jmeter/blob/trunk/src/components/org/apache/jmeter/visualizers/backend/graphite/PickleGraphiteMetricsSender.java
- https://github.com/BrightcoveOS/metrics-graphite-pickle/blob/master/src/main/java/com/brightcove/metrics/reporting/GraphitePickleReporter.java

**Comment by cyrille-leclerc** [comment]
Hi Guillaume,

For a good implementation of the pickle protocol in java, I would recommend
you to have a look at embedded-jmxtrans ;-)

I'm lost hiking in the French Alps, I have very limited internet
connectivity so I can't send you the link.

Cyrille

**Comment by gehel** [comment]
@cyrille-leclerc I had a look at [embeded-jmxtrans](https://github.com/jmxtrans/embedded-jmxtrans/blob/master/src/main/java/org/jmxtrans/embedded/output/GraphitePickleWriter.java) and was surprised that you use directly Jython. That's probably a robust solution, but it brings > 25Mo of additional dependencies. I expected this to be a no go for you ...

I'd like to have your reasoning once you are back from hiking. (Enjoy the vacation !)

**Comment by cyrille-leclerc** [comment]
I looked at the pickle protocol, it seems to be the Python equivalent of
Java serialization so I admitted that I touched my limit :-)

If you want pickle serialization in Java, it will cost you 25 MB :-)

I thought it would be cool if graphite did support a language agnostic
serialization protocol such as protocol buffer.

Cyrille

**Comment by gehel** [comment]
Well, the Graphite TCP protocol is simple enough that it does not really need a higher level abstraction ...

**Comment by gehel** [comment]
Just thinking, but to send data to Graphite in Pickle format, we only need a fairly small subset of Pickle working. It might be doable to implement just the minimal subset. I'm fairly sure I'm not going to have time to have a look any time soon, but if anyone wants to submit a PR...

**Comment by maauso** [comment]
Any update? we have thousands of servers with jmxtrans in separate services and it send hundred metrics for each server, then with the normal connection, the servers needs open one conection for each metrics.

**Comment by gehel** [comment]
No updates. This is a very, very low priority task for me. The Graphite line protocol can already handle a lot of metrics. And it does send multiple metrics over the same connection. If you see only one metric per connection on your setup, that's probably a bug (and I would be interested to know more!).

**Comment by cyrille-leclerc** [comment]
@maauso I have implemented the pickle writer a while ago on embedded-jmxtrans and on jmxtrans-agent. As @gehel have said, it is surprising that you report "servers needs open one connection for each metrics".

Regarding the performance boost of the pickle writer, I don't remember but I don't think that it is a 10x factor, I don't think it's a disruption.

**Comment by maauso** [comment]
Hi!

It's true, sorry, jmxtrans only use 1 connection for send all metrics.

@cyrille-leclerc If we want use pickle in jmxtrans-agent, we need change it

``` "outputWriters":[
                  {
                     "@class":"com.googlecode.jmxtrans.model.output.GraphiteWriter",
```
to 

``` "outputWriters":[
                  {
                     "@class":"com.googlecode.jmxtrans.model.output.GraphitePickleWriter,
```

Thanks!
